### PR TITLE
[NHC] Add support for verifying chain ID and role type

### PR DESCRIPTION
**Pull request by banool**
## Description
Currently it's possible for the user to pass in a URL to a testnet fullnode but select a devnet fullnode baseline (for example) and then run into potentially strange behaviour down the line. With this change, we explicitly check the chain ID and role type and return an evaluation if they don't match the expected baseline. This also adds most of the necessary functionality for an autodetect mode, where you don't have to pass in a baseline configuration at all and we just figure it out based on role type and chain ID.

This evaluator is enabled by default and cannot be disabled. This allows future evaluators to make certain assumptions.

## Test Plan
Generate appropriate config for testing a local fullnode. Notably we leave chain ID and role type unspecified here:
```
cargo run -- configuration create --configuration-name local_devnet_fullnode --configuration-name-pretty "Local Devnet Fullnode" --url http://localhost/ --evaluators state_sync_version > /tmp/local_devnet_fullnode.yaml
```

Run NHC:
```
cargo run -- server run --baseline-node-config-paths /tmp/local_devnet_fullnode.yaml
```

Hit it with a request (the target node == the baseline node here so, so we're comparing it to itself):
```
curl 'localhost:20121/check_node?node_url=http://localhost&baseline_configuration_name=local_devnet_fullnode'
```
```
{
  "evaluation_results": [
    {
      "headline": "Chain ID and role type reported by baseline and target match",
      "score": 100,
      "explanation": "The node under investigation reported the same chain ID (18) and role type (full_node) as is reported by the baseline node",
      "evaluator_name": "node_identity",
      "category": "node_identity",
      "links": []
    },
    {
      "headline": "State sync version is within tolerance",
      "score": 100,
      "explanation": "Successfully pulled metrics from target node twice, saw the version was progressing, and saw that it is within tolerance of the baseline node. Target version: 302087. Baseline version: 302087. Tolerance: 5000.",
      "evaluator_name": "state_sync_version",
      "category": "state_sync",
      "links": []
    }
  ],
  "summary_score": 100,
  "summary_explanation": "100: Awesome!"
}
```

See that if you configure a baseline to expect a particular chain ID / role type and the baseline has something else in reality, NHC will refuse to start up:
```
cargo run -- configuration create --configuration-name local_devnet_fullnode --configuration-name-pretty "Local Devnet Fullnode" --url http://localhost/ --evaluators state_sync_version > /tmp/local_devnet_fullnode.yaml --chain-id 2
```
```
$ cargo run -- server run --baseline-node-config-paths /tmp/local_devnet_fullnode.yaml
Error: Failed to build baseline node configurations

Caused by:
    0: Failed to fetch chain ID and role type for local_devnet_fullnode
    1: Chain ID mismatch: The baseline configuration local_devnet_fullnode says the chain ID is TESTNET but the node reports chain ID 18
```

---
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/1630)

### CODE REVIEW

File: ecosystem/node-checker/src/configuration/types.rs [hunk 1]
```diff
@@ -71,29 +76,46 @@ pub struct NodeConfiguration {
     pub runner_args: RunnerArgs,
 }
 
+// TODO: Having comments like "only call this after X" is obviously not great.
+// It'd be better to have an enum with two variants, e.g. unfetched and fetched.
```
Comment (JoshLind) [inline ecosystem/node-checker/src/configuration/types.rs:80]: +1

File: ecosystem/node-checker/src/evaluators/direct/node_identity.rs [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+pub struct NodeIdentityEvaluator {
+    args: NodeIdentityEvaluatorArgs,
+}
+
+impl NodeIdentityEvaluator {
+    pub fn new(args: NodeIdentityEvaluatorArgs) -> Self {
+        Self { args }
+    }
+
+    fn build_evaluation_result(
```
Comment (JoshLind) [inline ecosystem/node-checker/src/evaluators/direct/node_identity.rs:84]: nit: if we end up doing this a lot, we could create a new trait for it (e.g., default implementation).

File: ecosystem/node-checker/src/evaluators/direct/node_identity.rs [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+                "Chain ID and role type reported by baseline and target match".to_string(),
+                100,
+                format!(
+                    "The node under investigation reported the same chain ID \"{}\" \
+                and role type \"{}\" as is reported by the baseline node",
+                    target_chain_id, target_role_type
+                ),
+            )
+        } else {
+            self.build_evaluation_result(
```
Comment (JoshLind) [inline ecosystem/node-checker/src/evaluators/direct/node_identity.rs:126]: nit: should we be more explicit and check each independently? Just thinking about the users and also any UIs built on top of this 🤔

File: ecosystem/node-checker/src/runner/blocking_runner.rs [hunk 4]
```diff
@@ -75,8 +84,26 @@ impl<M: MetricCollector> BlockingRunner<M> {
...
+        };
+
+        debug!("Confirming node identity matches");
+        let node_identity_evaluations = self
+            .node_identity_evaluator
+            .evaluate(&direct_evaluator_input)
+            .await
+            .map_err(RunnerError::NodeIdentityEvaluatorError)?;
+
+        // Exit early if the node identity evaluation returned a non-passing result.
```
Comment (JoshLind) [inline ecosystem/node-checker/src/runner/blocking_runner.rs:102]: Nice!

File: ecosystem/node-checker/src/evaluators/direct/node_identity.rs [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    Ok((chain_id, role_type))
+}
+
+#[derive(Debug, ThisError)]
+pub enum NodeIdentityEvaluatorError {
+    #[error("Failed to get node identity from node: {0}")]
+    GetNodeIdentityError(anyhow::Error),
+}
+
+// TODO: Consider taking chain_id and role_type here instead.
```
Comment (JoshLind) [inline ecosystem/node-checker/src/evaluators/direct/node_identity.rs:69]: Yeah, I guess then we'd actually be able to write some unit tests for this as we wouldn't need to ping endpoint urls in the tests?

File: ecosystem/node-checker/src/server/run.rs [hunk 6]
```diff
@@ -54,28 +57,37 @@ pub struct Run {
...
         "Running with the following configuration: {:#?}",
         configurations_manager.configurations
     );
 
-    let target_metric_collector = match args.target_node_url {
-        Some(ref url) => Some(ReqwestMetricCollector::new(
-            url.clone(),
-            args.target_metrics_port,
-        )),
+    let preconfigured_test_node = match args.target_node_url {
```
Comment (JoshLind) [inline ecosystem/node-checker/src/server/run.rs:70]: nit: `args.target_node_url.map(|url| ...)`?

**Review by JoshLind — APPROVED** [review]
Looks reasonable to me!

File: ecosystem/node-checker/src/evaluators/direct/node_identity.rs (hunk 3 above)
Comment (banool) [reply ecosystem/node-checker/src/evaluators/direct/node_identity.rs:126]: Yeah sure seems like a good idea.

**Pull request merged**